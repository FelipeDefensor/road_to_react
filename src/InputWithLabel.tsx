import * as React from 'react'
import styles from "./App.module.css"

type InputWithLabelProps = {
  id: string,
  value: string,
  isFocused?: boolean,
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void,
  type?: string
  children: React.ReactNode
  }

export const InputWithLabel: React.FC<InputWithLabelProps> = ({
   id,
   value,
   isFocused,
   onInputChange,
   type='text',
   children
  }) => {

  return (
    <div>
      <label htmlFor={id} className={styles.label}>{children}</label>
      &nbsp;
      <input
        type={type}
        id={id}
        value={value}
        autoFocus={isFocused}
        onChange={onInputChange}
        className={styles.input}
      />
    </div>
  )
}