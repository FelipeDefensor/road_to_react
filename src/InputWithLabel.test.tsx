import * as React from 'react'

import { vi, describe, it, expect } from 'vitest'

import {
    fireEvent,
    render,
    screen
} from '@testing-library/react'

import { InputWithLabel } from './InputWithLabel'

describe('InputWithLabel', () => {

    const inputWithLabelProps = {
        id: 'test',
        value: 'initial',
        onInputChange: vi.fn(),
        children: <p>Dummy child</p>
    }


    it ('initial value is set correctly', () => {
        render(<InputWithLabel {...inputWithLabelProps }/>)

        expect(screen.getByDisplayValue(/initial/)).toBeInTheDocument()
    })

    it('displays children', () => {
        render(<InputWithLabel {...inputWithLabelProps}/>)

        expect(screen.getByText('Dummy child')).toBeInTheDocument()

        screen.debug()

    })

    it('calls onInputChange when input changes', async () => {
        render(<InputWithLabel {...inputWithLabelProps}/>)

        fireEvent.change(screen.getByRole('textbox'), {target:{value:'new value'}})


        expect(inputWithLabelProps.onInputChange).toBeCalledTimes(1)
        expect(inputWithLabelProps.onInputChange).toHaveBeenCalledWith(
        expect.objectContaining({
            target: expect.objectContaining({
                value: "initial" // not sure if this should be the value instead of 'new value'
            })
        })) 
    })

    it('matches snapshot', () => {
        const { container } = render(< InputWithLabel {...inputWithLabelProps} />)
        expect(container.firstChild).toMatchSnapshot()
    })
})