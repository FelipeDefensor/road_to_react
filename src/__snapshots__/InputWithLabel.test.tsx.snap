// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputWithLabel > matches snapshot 1`] = `
<div>
  <label
    class="_label_6dc87e"
    for="test"
  >
    <p>
      Dummy child
    </p>
  </label>
  Â 
  <input
    class="_input_6dc87e"
    id="test"
    type="text"
    value="initial"
  />
</div>
`;
